core/parentchain/block-import-queue/src/block_import_queue.rs

```diff
@@ -0,0 +1,144 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! Block import queue implementation
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	PopFromBlockQueue, PushToBlockQueue,
+};
+use std::{collections::VecDeque, vec::Vec};
+
+/// Block import queue.
+///
+/// Uses RwLock internally to guard against concurrent access and ensure all operations are atomic.
+pub struct BlockImportQueue<PB> {
+	queue: RwLock<VecDeque<PB>>,
+}
+
+impl<PB> BlockImportQueue<PB> {
+	pub fn is_empty(&self) -> Result<bool> {
+		let queue_lock = self.queue.read().map_err(|_| Error::PoisonedLock)?;
+		Ok(queue_lock.is_empty())
+	}
+}
+
+impl<PB> Default for BlockImportQueue<PB> {
+	fn default() -> Self {
+		BlockImportQueue { queue: Default::default() }
+	}
+}
+
+impl<PB> PushToBlockQueue for BlockImportQueue<PB> {
+	type BlockType = PB;
+
+	fn push_multiple(&self, blocks: Vec<PB>) -> Result<()> {
+		let mut queue_lock = self.queue.write().map_err(|_| Error::PoisonedLock)?;
+		queue_lock.extend(blocks);
+		Ok(())
+	}
+
+	fn push_single(&self, block: PB) -> Result<()> {
+		let mut queue_lock = self.queue.write().map_err(|_| Error::PoisonedLock)?;
+		queue_lock.push_back(block);
+		Ok(())
+	}
+}
+
+impl<PB> PopFromBlockQueue for BlockImportQueue<PB> {
```

Comment:
Why are you differentiating the traits for pop & push? Will there be ever a case where only one of them are needed?

Reply:
With trait mixing being so simple in Rust, I try to make my traits as granular as possible. Right now I don't have a use-case (yet) where I need only one. But it's imaginable, because the `pop` is only for retrieving, the `push` only for adding.

Reply:
I see, thanks for explaining :)